### ISSUE TITLE
Quarkus devmode does not warn when port is not available

### ISSUE THREAD

**Issue by glefloch** [quarkusio/quarkus#17950]
## Describe the bug

Starting two quarkus apps end up with an `ERROR` and a `WARNING` in the second one indicating that the port is already binded. 
When running a docker container on port 8080 and starting a quarkus app, this warning does not appear, everything looks good but the app is not accessible. 

cc @sebastienblanc I think this is the same error  you hit on your demo last week

### Expected behavior
Quarkus should display a warning such as: 

````
2021-06-15 22:22:40,108 ERROR [io.qua.run.Application] (Quarkus Main Thread) Port 8080 seems to be in use by another process. Quarkus may already be running or the port is used by another application.
2021-06-15 22:22:40,110 WARN  [io.qua.run.Application] (Quarkus Main Thread) Use 'netstat -anv | grep 8080' to identify the process occupying the port.
2021-06-15 22:22:40,110 WARN  [io.qua.run.Application] (Quarkus Main Thread) You can try to kill it with 'kill -9 <pid>'.
````

### Actual behavior
Quarkus does not display this error when the port is occupied by a docker container

## To Reproduce

1. Start a docker container that exposes something on port 8080
2. Start a quarkus app in dev mode with a simple rest resource
3. everything looks ok but, hitting http://localhost:8080 will hit the docker container

### Output of `uname -a` or `ver`
Darwin Guillaumes-MacBook-Pro.local 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:11 PDT 2020; root:xnu-7195.50.7~2/RELEASE_X86_64 x86_64

### Output of `java -version`
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06, mixed mode, sharing)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
It appears with both tools

## Additional context
(Add any other context about the problem here.)

**Comment by mkouba** [comment]
@glefloch So I've tried to reproduce this issue locally but the ERROR and WARNING messages were correctly displayed. I've tried to (1) start two apps in the dev mode and also (2) build the image first, then run the image, e.g. ` docker run -i --rm -p 8080:8080 mkouba/qute-quickstart:1.0.0-SNAPSHOT` and start the other app in the dev mode.

But it might be OS-related, I'm using Linux.

**Comment by mkouba** [comment]
Does it also occur if you run the container and the other app in the prod mode, i.e. `java -jar target/quarkus-app/quarkus-run.jar`?

**Comment by glefloch** [comment]
If I run the app in prod mode, the warning occurs but not in dev mode, I'm running on OSX: 

```
Listening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2021-06-22 13:01:36,994 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.257s. Listening on: http://localhost:8080
2021-06-22 13:01:37,007 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2021-06-22 13:01:37,007 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, rest-client, resteasy, resteasy-jackson, smallrye-context-propagation]
2021-06-22 13:01:37,007 INFO  [quarkus] (Quarkus Main Thread) Tests paused, press [r] to resume, [h] for more options>

<===========--> 88% EXECUTING [15s]
> :quarkusDev
                                                                                                                                                                                                                                                                                java -jar build/quarkus-app/quarkus-run.jar

__  ____  __  _____   ___  __ ____  ______
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2021-06-22 13:01:45,502 ERROR [io.qua.run.Application] (main) Port 8080 seems to be in use by another process. Quarkus may already be running or the port is used by another application.
2021-06-22 13:01:45,514 WARN  [io.qua.run.Application] (main) Use 'netstat -anv | grep 8080' to identify the process occupying the port.
2021-06-22 13:01:45,515 WARN  [io.qua.run.Application] (main) You can try to kill it with 'kill -9 <pid>'.
 ~/Downloads/code-with-quarkus 
```

This is really weird

**Comment by mkouba** [comment]
Hm,it seems that you're using gradle, right?

**Comment by glefloch** [comment]
yes but, I tested it with maven too, and I have the same problem.

**Comment by mkouba** [comment]
> yes but, I tested it with maven too, and I have the same problem.

It would be helpful if someone else running on OSX could test it as well. @maxandersen do you have a minute?

**Comment by mkouba** [comment]
We'll need someone with OSX to verify this. @holly-cummins @maxandersen do you have some spare cycles?

**Comment by maxandersen** [comment]
running 2 quarkus dev commands at the same time fails!

Its as if docker does not block listening. I guess because it might be on other interface?

**Comment by maxandersen** [comment]
Reverse fails same way - quarkus dev starts, docker starts too but this time quarkus dev wins.

My guess is that the port conflicts just aren't detectable in this case as treated as separate networks (my guess)

**Comment by maxandersen** [comment]
suspicion confirmed:

with `quarkus dev -Dquarkus.http.host=0.0.0.0` port conflicts gets detected.

but with 

`quarkus dev -Dquarkus.http.host=localhost` port conflict is not happening.

I kinda assumed 0.0.0.0 was the default but apparently not.

But that explains why. Not sure if any good fix for it.

**Comment by mkouba** [comment]
According to the [docs](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.host): _"The HTTP host In dev/test mode this defaults to localhost, in prod mode this defaults to 0.0.0.0"_.

**Comment by maxandersen** [comment]
Ah yes; to not expose everything.

But yeah - that explains why.

**Comment by geoand** [comment]
Is this still an issue with the latest versions of Quarkus?

**Comment by holly-cummins** [comment]
It is, sadly. If I do 

```
podman run -d -p 8080:80 --name test-server nginx:alpine

quarkus dev
```

it starts without error, and gives a cheerful confirmation that it's doing great:

```
2026-02-26 09:20:31,137 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.31.1) started in 9.446s. Listening on: http://localhost:8080
```

When I do `w` to bring up the application window, I get  

<img width="1388" height="731" alt="Image" src="https://github.com/user-attachments/assets/2cc6fe75-b796-4788-9f40-358670df5dfa" />

**Comment by geoand** [comment]
Thanks for checking!
